File: medium/python/1358.py
# https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/description/

class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        cnt = [0] * 3
        l, r = 0, 0
        res = 0

        while r < len(s):
            cnt[ord(s[r]) - ord('a')] += 1

            while cnt[ord(s[l]) - ord('a')] > 1:
                cnt[ord(s[l]) - ord('a')] -= 1
                l += 1

            if all(cnt):
                res += l + 1

            r += 1
            
        return res

File: medium/cpp/1358.cpp
// https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/

class Solution
{
public:
    int numberOfSubstrings(std::string s)
    {
        std::array<int, 3> cnt{ 0 };

        auto res = 0;

        for (auto l = 0, r = 0; r < std::size(s); r++)
        {
            cnt[s[r] - 'a']++;

            while (cnt[s[l] - 'a'] > 1)
            {
                cnt[s[l] - 'a']--;
                l++;
            }

            if (*std::min_element(std::begin(cnt), std::end(cnt)) == 1)
            {
                res += l + 1;
            }
        }

        return res;
    }
};